### PR TITLE
Update webpack 4 to webpack 5

### PR DESCRIPTION
**Pull request by rjchow**
- updated to webpack 5, along with required webpack plugins
- updated webpack config to match updated versions
- fixed dependency extraction config
- port unminify.js to webpack 5

Tested in post-merge repo structure, works fine but will need minor modifications to the webpack config path changes as the webpack config changed slightly.

Consumers of @woocommerce/dependency-extraction-webpack-plugin will have a breaking change:
- change webpack config output.libraryTarget from 'this' to 'window'

### Detailed test instructions:

- Basic smoke test just to check that everything still builds
- Check that `pnpm run build` and `pnpm run test:client` still work

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -8,7 +8,6 @@ const CustomTemplatedPathPlugin = require( '@wordpress/custom-templated-path-web
 const BundleAnalyzerPlugin = require( 'webpack-bundle-analyzer' )
 	.BundleAnalyzerPlugin;
 const MomentTimezoneDataPlugin = require( 'moment-timezone-data-webpack-plugin' );
-const TerserPlugin = require( 'terser-webpack-plugin' );
```
Comment (chihsuan) [inline webpack.config.js:11]: Looks like we don't use this in other places, shall we remove it from the `package.json`?

File: unminify.js [hunk 2]
```diff
@@ -8,19 +8,20 @@
  * 2. Remove check for development mode - we always want unminified files.
  * 3. Remove BannerPlugin support - we don't use it.
  * 4. Remove the 'min' suffix from the chunk loaded in the new `mainEntry` option.
- * 5. Hook into compilation later so we're running after Source Map generation.
+ * 5. Hook into compilation later so we're running after Source Map generation. (https://webpack.js.org/api/compilation-hooks/: PROCESS_ASSETS_STAGE_OPTIMIZE_INLINE)
```
Comment (chihsuan) [inline unminify.js:11]: @rjchow Sorry, could you share some details on what and why you change this file? I think prettier makes it harder to review this.

File: webpack.config.js [hunk 3]
```diff
@@ -142,13 +145,14 @@ const webpackConfig = {
...
-			wcAdminPackages.map( ( packageName ) => ( {
+		new CopyWebpackPlugin({
+
+			patterns: wcAdminPackages.map( ( packageName ) => ( {
 				from: `./packages/${ packageName }/build-style/*.css`,
-				to: `./${ packageName }/`,
-				flatten: true,
-				transform: ( content ) => content,
+				to: `./${ packageName }/[name][ext]`,
+				noErrorOnMissing: true
```
Comment (chihsuan) [inline webpack.config.js:153]: Curious, why do we need to set `noErrorOnMissing` to `true` here?

**Review by chihsuan — CHANGES_REQUESTED** [review]
@rjchow Tested well with instructions! I just have some questions about the changes. 

Besides, I notice that `node --trace-deprecation node_modules/webpack/bin/webpack.js` shows some `DeprecationWarning`. Could you take a look? Thanks!

File: unminify.js (hunk 2 above)
Comment (rjchow) [reply unminify.js:11]: 😆 It is hard to review regardless of prettier or not!

Basically I had to apply a 3-way "merge" between
1. Original author's webpack v4 version: https://github.com/leftstick/unminified-webpack-plugin/commit/734953e370fd55a3d68bfe3a1425adcaf8103d2b (deleted version of index.js)
2. Our webpack v4 version (previous commit of this file)
3. Original author's webpack v5 version: https://github.com/leftstick/unminified-webpack-plugin/blob/master/support5.js

I'm no expert at webpack so I largely relied on the heuristic of "is it working correctly yet?" to guide me through this.

When you look at the original author's webpack v5 version I think that accounts for most of the changes, then I applied the changes documented in this header comment block.

File: webpack.config.js (hunk 1 above)
Comment (rjchow) [reply webpack.config.js:11]: Thanks! I missed that out

File: webpack.config.js (hunk 3 above)
Comment (rjchow) [reply webpack.config.js:153]: Some packages don't have any css at all, so when this pattern match tries to run against such a folder it errors out

**Comment by rjchow** [comment]
> @rjchow Tested well with instructions! I just have some questions about the changes.
> 
> Besides, I notice that `node --trace-deprecation node_modules/webpack/bin/webpack.js` shows some `DeprecationWarning`. Could you take a look? Thanks!

It looks like one of those is from @wordpress/custom-templated-path-webpack-plugin, and 3 of those are from webpack-rtl-plugin https://github.com/romainberger/webpack-rtl-plugin/issues/34

They all still exist in the latest versions :(

**Comment by adrianduffell** [comment]
Thank you @rjchow, great job wrangling this.

> Consumers of @woocommerce/dependency-extraction-webpack-plugin will have a breaking change:

> change webpack config output.libraryTarget from 'this' to 'window'

Thanks for highlighting the breaking change!

**Comment by rjchow** [comment]
Shout out to @joelclimbsthings for laying the groundwork in https://github.com/woocommerce/woocommerce-admin/issues/8326 😁

File: unminify.js (hunk 2 above)
Comment (chihsuan) [reply unminify.js:11]: > When you look at the original author's webpack v5 version I think that accounts for most of the changes, then I applied the changes documented in this header comment block.

Thanks @rjchow. This makes sense to me!

**Comment by chihsuan** [comment]
@rjchow The changes look good to me! I just found `pnpm run storybook` is not working now. 😞 Should we also update the [storybook's webpack config](https://github.com/woocommerce/woocommerce-admin/blob/fix/webpack5/storybook/webpack.config.js)?

Screen Shot 2022-03-18 at 09 52 05

**Comment by rjchow** [comment]
> @rjchow The changes look good to me! I just found `pnpm run storybook` is not working now. 😞 Should we also update the [storybook's webpack config](https://github.com/woocommerce/woocommerce-admin/blob/fix/webpack5/storybook/webpack.config.js)?
> 
> Screen Shot 2022-03-18 at 09 52 05

Aye I forgot about that, thanks for catching it! I'll fix that

**Comment by rjchow** [comment]
@chihsuan Done! 😃

**Review by chihsuan — APPROVED** [review]
Tested well and LGTM. Way to go! 🚀

**Pull request merged**